File: data_structures/linked_list/singly_linked_list.go
package main

import "fmt"

type Node struct {
	Data interface{}
	Next *Node
}

type LinkedList struct {
	Head *Node
	Size int
}

func (ll *LinkedList) InsertHead(data interface{}) {
	newNode := &Node{Data: data}
	newNode.Next = ll.Head
	ll.Head = newNode
	ll.Size++
}

func (ll *LinkedList) InsertTail(data interface{}) {
	newNode := &Node{Data: data}
	if ll.Head == nil {
		ll.Head = newNode
	} else {
		current := ll.Head
		for current.Next != nil {
			current = current.Next
		}
		current.Next = newNode
	}
	ll.Size++
}

func (ll *LinkedList) DeleteHead() interface{} {
	if ll.Head == nil {
		return nil
	}
	data := ll.Head.Data
	ll.Head = ll.Head.Next
	ll.Size--
	return data
}

func (ll *LinkedList) Reverse() {
	var prev *Node = nil
	current := ll.Head
	for current != nil {
		next := current.Next
		current.Next = prev
		prev = current
		current = next
	}
	ll.Head = prev
}

func (ll *LinkedList) PrintList() {
	current := ll.Head
	for current != nil {
		fmt.Printf("%v -> ", current.Data)
		current = current.Next
	}
	fmt.Println("nil")
}

func main() {
	ll := &LinkedList{}
	ll.InsertTail(1)
	ll.InsertTail(2)
	ll.InsertTail(3)
	ll.PrintList()
	
	ll.InsertHead(0)
	ll.PrintList()
	
	ll.Reverse()
	ll.PrintList()
	
	fmt.Println("Deleted:", ll.DeleteHead())
	ll.PrintList()
}

File: data_structures/linked_list/singly_linked_list.rs
#[derive(Debug, Clone)]
struct Node<T> {
    data: T,
    next: Option<Box<Node<T>>>,
}

pub struct LinkedList<T> {
    head: Option<Box<Node<T>>>,
    size: usize,
}

impl<T: std::fmt::Debug + Clone> LinkedList<T> {
    pub fn new() -> Self {
        LinkedList { head: None, size: 0 }
    }

    pub fn is_empty(&self) -> bool {
        self.head.is_none()
    }

    pub fn len(&self) -> usize {
        self.size
    }

    pub fn insert_head(&mut self, data: T) {
        let new_node = Box::new(Node {
            data,
            next: self.head.take(),
        });
        self.head = Some(new_node);
        self.size += 1;
    }

    pub fn insert_tail(&mut self, data: T) {
        let new_node = Box::new(Node {
            data,
            next: None,
        });

        if self.head.is_none() {
            self.head = Some(new_node);
        } else {
            let mut current = self.head.as_mut().unwrap();
            while current.next.is_some() {
                current = current.next.as_mut().unwrap();
            }
            current.next = Some(new_node);
        }
        self.size += 1;
    }

    pub fn delete_head(&mut self) -> Option<T> {
        self.head.take().map(|node| {
            self.head = node.next;
            self.size -= 1;
            node.data
        })
    }
    
    pub fn reverse(&mut self) {
        let mut prev = None;
        let mut current = self.head.take();
        
        while let Some(mut node) = current {
            let next = node.next.take();
            node.next = prev;
            prev = Some(node);
            current = next;
        }
        self.head = prev;
    }

    pub fn print_list(&self) {
        let mut current = &self.head;
        while let Some(node) = current {
            print!("{:?} -> ", node.data);
            current = &node.next;
        }
        println!("None");
    }
}

fn main() {
    let mut ll = LinkedList::new();
    ll.insert_tail(1);
    ll.insert_tail(2);
    ll.insert_tail(3);
    ll.print_list();

    ll.insert_head(0);
    ll.print_list();
    
    ll.reverse();
    ll.print_list();
    
    println!("Deleted: {:?}", ll.delete_head());
    ll.print_list();
}
